### PR TITLE
Fix issue #5365: Support for Sparse MoE models like Camelidae and Sparsetral

### PR DESCRIPTION
**Pull request by nganhuongg**
## Overview

This PR adds initial support for the Sparsetral sparse-adapter MoE architecture described in #5365.

The scope of this PR is intentionally limited to **Sparsetral**. Sparsetral differs from conventional full-MoE models such as Mixtral. It keeps the dense Mistral FFN active and adds routed lightweight adapter experts on top of the dense FFN output. Because of this, it cannot use the existing assumption that any model with `expert_count > 0` contains replacement-style full FFN experts.

This PR adds the required support across conversion, GGUF metadata, tensor loading, graph construction, and quantization:

- registers `modeling_sparsetral.MistralForCausalLM` in the Hugging Face-to-GGUF conversion pipeline;
- reuses the existing LLaMA/Mistral conversion path for the dense backbone;
- maps Sparsetral-specific configuration fields:
  - `num_experts`;
  - `topk`;
  - `adapter_dim`;
  - softmax router scoring;
- adds dedicated GGUF tensor names for:
  - the sparse-adapter router;
  - expert-stacked `adapter_down` weights;
  - expert-stacked `adapter_up` weights;
- buffers and stacks individual Hugging Face adapter expert tensors into 3D GGUF tensors;
- adds adapter bottleneck metadata required by the C++ loader;
- adds a sparse-adapter loading path that loads both:
  - the normal dense FFN tensors;
  - the routed sparse-adapter tensors;
- reuses the existing `build_ffn()`, `build_lora_mm()`, and `build_moe_ffn()` infrastructure to construct the Sparsetral inference graph;
- computes routing from the normalized FFN input;

The implemented FFN path is conceptually:

```text
ffn_norm      = RMSNorm(ffn_input)
dense_out     = dense_ffn(ffn_norm)
router_logits = adapter_router(ffn_norm)
adapter_out   = routed_adapter_moe(dense_out, router_logits)
ffn_out       = dense_out + adapter_out
layer_out     = ffn_input + ffn_out
```

### Conversion and loading validation

The full Sparsetral checkpoint was converted successfully:

```text
expert count = 16
experts used count = 4
expert score gating function = softmax
n_tensors = 387
total_size = 18.8G
Model successfully exported
```

The converted GGUF loads successfully without entering the existing full-MoE loading path or requesting Mixtral-style tensors such as:

```text
ffn_gate_inp
ffn_gate_exps
ffn_up_exps
ffn_down_exps
```

Instead, the loader reads the dense FFN and sparse-adapter tensors:

```text
ffn_gate
ffn_up
ffn_down
ffn_moe_adapter_gate
ffn_moe_adapter_down_exps
ffn_moe_adapter_up_exps
```

### Graph validation

The sparse-adapter graph branch was inspected with `llama-debug`:

```powershell
.\build-amd64-test\bin\llama-debug.exe `
    -m .\sparsetral-new.gguf `
    -p "Hello" `
    -n 1 `
    -ngl 0 `
    --verbose `
    --tensor-filter "ffn_moe_adapter"
```

The executed graph contains:

```text
ffn_moe_adapter_logits
ffn_moe_adapter_out
```

This confirms that the adapter tensors are connected to and evaluated by the computation graph.

### Numerical validation

I compared the final next-token logits against the Hugging Face Sparsetral reference implementation using the same prompt and tokenizer. The converted GGUF logits were generated locally with `llama-debug`, and the outputs were compared using llama.cpp's causal logit comparison script.

The top logits and predicted tokens were closely aligned between the Hugging Face and llama.cpp implementations.

### Regression validation

I tested the existing dense path using:

```text
stories15M-q4_0.gguf
```

The model loaded and generated without requesting sparse-adapter or full-MoE tensors.

I also tested the existing full-MoE path using:

```text
allenai/OLMoE-1B-7B-0125-GGUF:Q2_K
```

The model loaded and generated successfully, confirming that the new sparse-adapter branch does not replace or break the existing full-MoE path.

The following focused regression tests passed:

```powershell
ctest `
    --test-dir build-amd64-test `
    --output-on-failure `
    -R "test-gguf|test-model-load-cancel|test-backend-sampler|test-save-load-state|test-thread-safety"
```

Result:

```text
7/7 tests passed
```

### Quantization validation

The converted Sparsetral GGUF was accepted by `llama-quantize` for:

```text
Q8_0
Q4_K_M
```

This verifies that the new expert-stacked adapter tensors are recognized by the quantization pipeline and retain valid tensor shapes.

## Requirements

- I have read and agree with the [contributing guidelines](https://github.com/ggml-org/llama.cpp/blob/master/CONTRIBUTING.md)

- AI usage disclosure: **YES** — AI was used to help navigate and understand the codebase, explain existing GGML and llama.cpp mechanisms, and assist with debugging and documentation. I manually reviewed the implementation and am responsible for every submitted code change.

### CODE REVIEW

**Comment by nganhuongg** [comment]
Hi @ggerganov, this is my first contribution to this project. I've addressed issue #5365 by adding initial Sparsetral support, including HF-to-GGUF conversion, sparse-adapter tensor loading, routed adapter graph execution, and quantization compatibility. Would appreciate a review when you have time!

**Comment by nganhuongg** [comment]
@ggml-bot Hi!

I used AI only as a learning aid to help me understand parts of the existing codebase and GGML concepts. All code changes were implemented by me, and I wrote the commit messages and PR description myself. I have manually reviewed every submitted change and take full responsibility for the contribution.
